File: backend/producers_test.go
package backend

import (
	"reflect"
	"sort"
	"testing"
)

func TestMatchPatterns(t *testing.T) {
	// shared pattern set: metro boomin with restricted "Metro" alias, wheezy with
	// long alias, and a producer whose name is a substring of unrelated words.
	patterns := []Pattern{
		{Term: "metro boomin", ProducerID: 1, IsAlias: false},
		{Term: "metro", ProducerID: 1, IsAlias: true, AliasArtistIDs: []int{100}},
		{Term: "wheezy", ProducerID: 2, IsAlias: false},
		{Term: "wheezy outta here", ProducerID: 2, IsAlias: true},
		{Term: "cash", ProducerID: 3, IsAlias: false},
	}

	tests := []struct {
		name          string
		filename      string
		songArtistIDs []int
		want          []int
	}{
		{
			name:     "exact producer name match",
			filename: "Song [Metro Boomin].mp3",
			want:     []int{1},
		},
		{
			name:          "alias match with valid artist",
			filename:      "Song [Metro].mp3",
			songArtistIDs: []int{100},
			want:          []int{1},
		},
		{
			name:          "restricted alias skipped when artist not in context",
			filename:      "Song [Metro].mp3",
			songArtistIDs: []int{999},
			want:          []int{},
		},
		{
			name:     "restricted alias skipped when no song artists",
			filename: "Song [Metro].mp3",
			want:     []int{},
		},
		{
			name:     "multi-word term does not match inside concatenated token",
			filename: "metroboomintrack.mp3", // no spaces, so multi-word "metro boomin" cannot appear
			want:     []int{},
		},
		{
			name:     "longer alias wins over substring",
			filename: "Track (Wheezy Outta Here).mp3",
			want:     []int{2},
		},
		{
			name:          "multiple producers in filename",
			filename:      "Song [Metro] (Wheezy Outta Here).mp3",
			songArtistIDs: []int{100},
			want:          []int{1, 2},
		},
		{
			name:     "snapshot: real filename matches both producers via names",
			filename: "Metro Boomin x Wheezy.mp3",
			want:     []int{1, 2},
		},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			got := MatchPatterns(tc.filename, patterns, tc.songArtistIDs)
			sort.Ints(got)
			want := append([]int{}, tc.want...)
			sort.Ints(want)
			if !reflect.DeepEqual(got, want) {
				t.Fatalf("MatchPatterns(%q, %v) = %v, want %v", tc.filename, tc.songArtistIDs, got, want)
			}
		})
	}
}

func TestMatchProducersFromFilenameHonorsAliasRestrictionsAndLongestMatches(t *testing.T) {
	app := newTestApp(t)

	restrictedArtist, err := app.CreateArtist(CreateArtistInput{Name: "Restricted Artist"})
	if err != nil {
		t.Fatalf("CreateArtist returned error: %v", err)
	}
	otherArtist, err := app.CreateArtist(CreateArtistInput{Name: "Other Artist"})
	if err != nil {
		t.Fatalf("CreateArtist returned error: %v", err)
	}

	metro, err := app.CreateProducerWithAliases(CreateProducerInput{
		Name: "Metro Boomin",
		Aliases: []AliasInput{
			{Name: "Metro", ArtistIDs: []int{restrictedArtist.ID}},
		},
	})
	if err != nil {
		t.Fatalf("CreateProducerWithAliases returned error: %v", err)
	}
	wheezy, err := app.CreateProducerWithAliases(CreateProducerInput{
		Name: "Wheezy",
		Aliases: []AliasInput{
			{Name: "Wheezy Outta Here"},
		},
	})
	if err != nil {
		t.Fatalf("CreateProducerWithAliases returned error: %v", err)
	}

	matches, err := app.MatchProducersFromFilename("Song [Metro] (Wheezy Outta Here).mp3", []int{restrictedArtist.ID})
	if err != nil {
		t.Fatalf("MatchProducersFromFilename returned error: %v", err)
	}
	if len(matches) != 2 {
		t.Fatalf("expected 2 producer matches, got %d", len(matches))
	}

	restrictedMatches, err := app.MatchProducersFromFilename("Song [Metro].mp3", []int{otherArtist.ID})
	if err != nil {
		t.Fatalf("MatchProducersFromFilename returned error: %v", err)
	}
	if len(restrictedMatches) != 0 {
		t.Fatalf("expected restricted alias to be ignored, got %v", restrictedMatches)
	}

	nameMatch, err := app.MatchProducersFromFilename("Metro Boomin x Wheezy.mp3", []int{restrictedArtist.ID})
	if err != nil {
		t.Fatalf("MatchProducersFromFilename returned error: %v", err)
	}
	if len(nameMatch) != 2 {
		t.Fatalf("expected direct producer names to match once each, got %v", nameMatch)
	}

	if metro.ID == wheezy.ID {
		t.Fatal("expected distinct producer ids")
	}
}

File: electron/domain/producers.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import Database from 'better-sqlite3';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import { migrate } from '../db/migrate';
import { registerLower } from '../db/connection';
import {
	matchPatterns,
	matchProducersFromFilename,
	createProducerWithAliases,
	type Pattern
} from './producers';
import { createArtist } from './artists';

const MIGRATIONS_DIR = join(dirname(fileURLToPath(import.meta.url)), '../../backend/migrations');

function freshDb(): Database.Database {
	const db = new Database(':memory:');
	registerLower(db);
	migrate(db, MIGRATIONS_DIR);
	return db;
}

// Parity mirror of backend/producers_test.go TestMatchPatterns — pure algorithm.
describe('matchPatterns', () => {
	const patterns: Pattern[] = [
		{ term: 'metro boomin', producerId: 1, isAlias: false },
		{ term: 'metro', producerId: 1, isAlias: true, aliasArtistIds: [100] },
		{ term: 'wheezy', producerId: 2, isAlias: false },
		{ term: 'wheezy outta here', producerId: 2, isAlias: true },
		{ term: 'cash', producerId: 3, isAlias: false }
	];

	const cases: { name: string; filename: string; songArtistIds?: number[]; want: number[] }[] = [
		{ name: 'exact producer name match', filename: 'Song [Metro Boomin].mp3', want: [1] },
		{
			name: 'alias match with valid artist',
			filename: 'Song [Metro].mp3',
			songArtistIds: [100],
			want: [1]
		},
		{
			name: 'restricted alias skipped when artist not in context',
			filename: 'Song [Metro].mp3',
			songArtistIds: [999],
			want: []
		},
		{ name: 'restricted alias skipped when no song artists', filename: 'Song [Metro].mp3', want: [] },
		{
			name: 'multi-word term does not match inside concatenated token',
			filename: 'metroboomintrack.mp3',
			want: []
		},
		{ name: 'longer alias wins over substring', filename: 'Track (Wheezy Outta Here).mp3', want: [2] },
		{
			name: 'multiple producers in filename',
			filename: 'Song [Metro] (Wheezy Outta Here).mp3',
			songArtistIds: [100],
			want: [1, 2]
		},
		{ name: 'real filename matches both via names', filename: 'Metro Boomin x Wheezy.mp3', want: [1, 2] }
	];

	for (const tc of cases) {
		it(tc.name, () => {
			expect(matchPatterns(tc.filename, patterns, tc.songArtistIds ?? [])).toEqual(tc.want);
		});
	}
});

// Parity mirror of TestMatchProducersFromFilenameHonorsAliasRestrictionsAndLongestMatches.
describe('matchProducersFromFilename', () => {
	let db: Database.Database;
	beforeEach(() => {
		db = freshDb();
	});

	it('honors alias restrictions and longest matches', () => {
		const restricted = createArtist(db, { name: 'Restricted Artist' });
		const other = createArtist(db, { name: 'Other Artist' });

		const metro = createProducerWithAliases(db, {
			name: 'Metro Boomin',
			aliases: [{ name: 'Metro', artistIds: [restricted.id] }]
		});
		const wheezy = createProducerWithAliases(db, {
			name: 'Wheezy',
			aliases: [{ name: 'Wheezy Outta Here', artistIds: [] }]
		});
		expect(metro.id).not.toBe(wheezy.id);

		expect(
			matchProducersFromFilename(db, 'Song [Metro] (Wheezy Outta Here).mp3', [restricted.id])
		).toHaveLength(2);

		expect(matchProducersFromFilename(db, 'Song [Metro].mp3', [other.id])).toHaveLength(0);

		expect(matchProducersFromFilename(db, 'Metro Boomin x Wheezy.mp3', [restricted.id])).toHaveLength(
			2
		);
	});

	it('rejects a duplicate alias across producers (case-insensitive)', () => {
		createProducerWithAliases(db, { name: 'A', aliases: [{ name: 'Dupe', artistIds: [] }] });
		expect(() =>
			createProducerWithAliases(db, { name: 'B', aliases: [{ name: 'dupe', artistIds: [] }] })
		).toThrow();
	});
});
